src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -95,8 +95,19 @@ impl<'a, 'gcx, 'tcx> Borrows<'a, 'gcx, 'tcx> {
                             rvalue: &mir::Rvalue<'tcx>,
                             location: mir::Location) {
                 if let mir::Rvalue::Ref(region, kind, ref lvalue) = *rvalue {
+                    let mut lvalue = lvalue;
+
+                    // `&a[i][j]` borrows entire `a`
+                    while let mir::Lvalue::Projection(ref proj) = *lvalue {
+                        if let mir::ProjectionElem::Index(_) = proj.elem {
```

Comment:
Don't we need to recur on other variants of `ProjectionElem` as well here? E.g. what if we there is a Field-access or a Downcast mixed into the projection?

Reply:
Thinking more about it, the real bug here may be the use of `Lvalue` within the `struct BorrowData`. It probably needs to be an abstraction of `Lvalue`, in order to represent cases like this while still allowing us to differentiate cases like `a[0].foo` and `a[0].bar`...

Reply:
(...having said that, my early attempt to make a concrete example illustrating a case where this PR hypothetically goes wrong is not actually showing a problem. So maybe *I* have misunderstood something here...)

Reply:
Here is a concrete example that this PR does not currently handle:

```rust
    struct S<T> { x: T, y: T, }
    let mut v = [S { x: 1, y: 2 },
                 S { x: 3, y: 4 },
                 S { x: 5, y: 6 }];
    let p = &v[0].x;
    if true {
        use_x(*p);
    } else {
        use_x(22);
    }
    v[0].x += 1; //[ast]~ ERROR cannot assign to `v[..]`
```

Reply:
Note that when I said "It probably needs to be an abstraction of Lvalue", I was basically talking about what the move paths code does; see here: https://github.com/rust-lang/rust/blob/master/src/librustc_mir/dataflow/move_paths/abs_domain.rs